src/Compiler/Utilities/Activity.fs

```diff
@@ -23,3 +57,78 @@ module Activity =
         activity
 
     let startNoTags (name: string) : IDisposable = activitySource.StartActivity(name)
+
+    let private escapeStringForCsv (o: obj) =
+        if isNull o then
+            ""
+        else
+            let mutable txtVal = o.ToString()
+            let hasComma = txtVal.IndexOf(',') > -1
+            let hasQuote = txtVal.IndexOf('"') > -1
+
+            if hasQuote then
+                txtVal <- txtVal.Replace("\"", "\\\"")
+
+            if hasQuote || hasComma then
+                "\"" + txtVal + "\""
+            else
+                txtVal
+
+    let private createCsvRow (a: Activity) =
+        let sb = new StringBuilder(128)
+
+        let appendWithLeadingComma (s: string) =
+            sb.Append(',') |> ignore
+            sb.Append(s) |> ignore
+
+        // "Name,StartTime,EndTime,Duration,Id,ParentId"
+        sb.Append(a.DisplayName) |> ignore
+        appendWithLeadingComma (a.StartTimeUtc.ToString("HH-mm-ss.ffff"))
+        appendWithLeadingComma ((a.StartTimeUtc + a.Duration).ToString("HH-mm-ss.ffff"))
+        appendWithLeadingComma (a.Duration.TotalSeconds.ToString("000.0000"))
+        appendWithLeadingComma (a.Id)
+        appendWithLeadingComma (a.ParentId)
+
+        let rec rootID (act: Activity) =
+            if isNull act.ParentId then act.Id else rootID act.Parent
+
+        appendWithLeadingComma (rootID a)
+
+        Tags.AllKnownTags
+        |> Array.iter (fun t -> a.GetTagItem(t) |> escapeStringForCsv |> appendWithLeadingComma)
+
+        sb.ToString()
+
+    let addCsvFileListener pathToFile =
+        if pathToFile |> File.Exists |> not then
+            File.WriteAllLines(
+                pathToFile,
+                [
+                    "Name,StartTime,EndTime,Duration(s),Id,ParentId,RootId,"
+                    + String.concat "," Tags.AllKnownTags
+                ]
+            )
+
+        let messages = new BlockingCollection<string>(new ConcurrentQueue<string>())
+
+        let l =
+            new ActivityListener(
+                ShouldListenTo = (fun a -> a.Name = activitySourceName),
+                Sample = (fun _ -> ActivitySamplingResult.AllData),
+                ActivityStopped = (fun a -> messages.Add(createCsvRow a))
+            )
+
+        ActivitySource.AddActivityListener(l)
+
+        let writerTask =
+            Task.Factory.StartNew(fun () ->
```

Comment:
Can we use backgroundTask/task CE here?

Reply:
It becomes this. I do not see it as a big win TBH (when it comes to communicating intent, they feel rather the same to me)

```
            let writerTask = backgroundTask {
                use sw = new StreamWriter(path = pathToFile, append = true)
                for msg in messages.GetConsumingEnumerable() do
                    sw.WriteLine(msg)
                } 
```

Reply:
It's just a matter of us eating our own dogfood. I.e. using our corelib's features where suits.

Reply:
Ok, that explanation makes sense. Will change and retest in a concurrent scenario.